xarray/tests/test_dataarray.py

```diff
@@ -1506,6 +1506,27 @@ def test_rename(self):
         renamed_kwargs = self.dv.x.rename(x="z").rename("z")
         assert_identical(renamed, renamed_kwargs)
 
+    def test_init_value(self):
```

Comment:
Could we add one more test for a 0-d array, given we have that code path? Then I think we're good to go!

Reply:
I'm relying on the the raise of no matching dimensions since I'm not adding a custom exception when a 0-d is passed.
I found that when a scalar array is passed and `coords` is a list, the error raised is self explanatory
```python
>>> xr.DataArray(np.array(1.0), coords=[np.arange(3), ['a', 'b']])
...
ValueError: coords is not dict-like, but it has 2 items, which does not match the 0 dimensions of the data
```

But when `coords` is given as a dictionary you don't get enough feedback

```python
>>> xr.DataArray(np.array(1), coords={'x': np.arange(4), 'y': 'a'}, dims=['x'])
...
KeyError: 'x'
```

The problem starts here 
https://github.com/pydata/xarray/blob/131f6022763d35edf461b11857d7c4ec6630b19d/xarray/core/dataarray.py#L123-L125

I was messing around and wrote a quick fix for it

```python
    if utils.is_dict_like(coords):
        for k, v in coords.items():
            new_coords[k] = as_variable(v, name=k)
        if shape == () and not any(
            v.shape == () for k, v in new_coords.items() if k in dims
        ):
            shape = (0,)
```

After which the error raised change to this
```python
>>> xr.DataArray(np.array(1), coords={'x': np.arange(4), 'y': 'a'}, dims=['x'])

ValueError: conflicting sizes for dimension 'x': length 0 on the data but length 4 on coordinate 'x'
```
Not sure if this is the most elegant solution but I need something along these lines to write a test for the 0-d

Reply:
Having a better error message would be good.

That said, we can add a test that 0d fails now - right? Or am I misunderstanding you?

(also fine to change the error message; but no need to delay this PR for that)

Reply:
So would it be wise to open an issue or another PR about the error message and leave this one as it is?
The test is written for the current behaviour that raises a `KeyError` instead of `ValueError` with the info about conflicting dimension size.

Reply:
I'm biased towards merging smaller PR then iterating, though ofc feel free to add if you wish